WT-10003 Add testutil APIs for checkpoint/flush_tier threads, and fix test/checkpoint accordingly.

ddanderson opened the pull request.

ddanderson commented:
@keitharnoldsmith and @raviprakashgiri29, this is ready for review.  There are a few new testutil APIs to allow us to have testutil control the timing of when flush_tier should happen, but yet allow the test application to actually make the calls to checkpoint and flush_tier.  This is needed because there is a lot of variation of checkpoint threads - both in the config passed into checkpoint and what else happens in these loops.  In terms of predictable replay, we'll need on the flushing node to track and record when flush_tier has happened, the `testutil_tiered_flush_complete` API does that.

Rather than write an extensive example of how this is used, I modified test/checkpoint to use it.  But I tried to do this minimally.  I did not do the full refactoring of getopt options in test/checkpoint, that will wait for another ticket.

keitharnoldsmith commented on test/checkpoint/checkpointer.c:
```diff
@@ -212,34 +165,42 @@ real_checkpointer(void)
+      flush_tier_config, sizeof(flush_tier_config), "flush_tier=(enabled,force),%s", ts_config));
+
+    /* Set up a random delay up to 5000 microseconds for the next flush. */
+    g.opts.tiered_flush_interval_us = __wt_random(&rnd) % 5001;
```
There are two places we do this initialization. Maybe factor it into a helper function/macro so there's just one place to update if we ever want to change the delay value.

raviprakashgiri29 replied:
Yeah, I agree with @keitharnoldsmith here. Calling a helper function to set the time interval would be more readable.

ddanderson replied:
Done.

keitharnoldsmith commented on test/checkpoint/checkpointer.c:
```diff
@@ -283,11 +258,14 @@ real_checkpointer(void)
-            delay = __wt_random(&rnd) % 5;
-            __wt_sleep(delay + 4, 0);
-        }
+
+        if (g.sweep_stress)
+            /* Random value between 4 and 8 seconds. */
+            delay = __wt_random(&rnd) % 5 + 4;
+        else
+            /* Just find out if we should flush_tier. */
+            delay = 0;
+        testutil_tiered_sleep(&g.opts, delay, &flush_tier);
```
I don't understand the intent of `sweep_stress`. Is it important that we wait 4-8 seconds between checkpoints? Because if tiered storage is enabled, we will never sleep that long because of the much shorter delay between flushes.

ddanderson replied:
Yes, I agree this should be changed.  I think `sweep_stress` means longer time (multiple seconds) between checkpoints, so each checkpoint is more stressful.  If that variable is not on, then we checkpoint in a tight loop.  This is unchanged behavior from the previous version (although the code got moved around).  The previous version also had a flush_tier thread that sleeps < 10000 usec.  Now, the flush is done by calling checkpoint("flush_tier=(enabled)") and that incorporates a checkpoint, of course.

I think the right thing, and in line with original intention is to keep this code (and thus the checkpoint interval behavior) the same, but make the flush_tier interval also look at `g.sweep_stress`.  If stress is off, interval is < 10000 usecs, as we have now. If stress is on, flush_tier interval is, say, between 5 and 15 seconds, that is, generally (but not always) somewhat longer than the checkpoint rate.

There is a question is how long the runs in test/checkpoint last (a 10 second run would not give us many flush calls).  It does not run on a timer, but rather completes when each of _W_ workers finishes _n_ operations.  For evergreen runs with `-s 1` (that sets `g.sweep_stress`), _W_ is 3 and _n_ is 100000.  I observe this runs ~45 seconds when run with timestamps, and ~90 seconds without timestamps.  So a flush_tier between 5 and 15 seconds should give something like 4 to 5 flush_tiers for these stress tests with timestamps, and ~9 flush_tiers without timestamps.  So that feels adequate, although when processors get faster, we may need to recalibrate (perhaps by doing more operations per run).

@sueloverso I expect you know test/checkpoint pretty well, can you comment on the above - I can brief you on what we're doing if it helps.

ddanderson commented on test/utility/tiered.c:
```diff
@@ -0,0 +1,111 @@
+    testutil_assert(opts->conn != NULL);
+
+    if (opts->tiered_storage && opts->tiered_flush_interval_us != 0) {
+        now = time_us();
+        opts->tiered_flush_next_us = now + opts->tiered_flush_interval_us;
```
Fixed.

ddanderson commented on test/checkpoint/checkpointer.c:
```diff
@@ -261,12 +222,26 @@ real_checkpointer(void)
 
-        if (!g.running)
+            /* Set up a random delay up to 5000 microseconds for the next flush. */
+            g.opts.tiered_flush_interval_us = __wt_random(&rnd) % 5001;
```
This change was somewhat intentional, but not completely well thought out.  Instead of (original) 5000 - 10000 or my first change (0 - 5000) I'm going to make it a random interval between (0- 10000).  I agree that it _shouldn't_ matter, but maybe (an occasional) very quick interval between flush_tier calls will stress something.  And having potential intervals up to 10000 means that we'll get some similar coverage as the original.

ddanderson commented:
@sueloverso, can you take a look at these changes.  In particular, I've changed the timing in stress mode.  Perhaps this wasn't running before with both flush_tier and "-s 1" stress mode, but if run that way, it would have checkpointed every 4 - 8 seconds and done a flush_tier every 5000 to 10000 usecs (i.e. orders of magnitude more often).  So I've adjusted the flush_tier rate in stress mode.

sueloverso commented on test/utility/tiered.c:
```diff
@@ -0,0 +1,109 @@
+
+    testutil_assert_errno(gettimeofday(&tv, NULL) == 0);
+    return ((uint64_t)tv.tv_sec * 1000000 + (uint64_t)tv.tv_usec);
+}
```
`tiered.c` seems an unusual place for this function. Maybe `test/utility/misc.c`?

Also this test program uses other wiredtiger functions, what about just calling `__wt_epoch` for all the clock portability support it already has?

ddanderson replied:
Fixed.  `__wt_epoch` needs a session, which required adding a session to a couple of the new testutil APIs.  But worth it for portability.

sueloverso reviewed (approved):
LGTM - I have one comment about the clock code.

keitharnoldsmith reviewed (approved):
LGTM

raviprakashgiri29 reviewed (approved):
Changes LGTM

ddanderson commented:
evergreen merge
